### PR TITLE
[master] Update hosted and BYOC agents used in our builds

### PR DESCRIPTION
**Pull request by dougbu**
- dotnet/aspnetcore-internal#3540

### CODE REVIEW

**Comment by dougbu** [comment]
Failing builds are kind-of ugly and happen way early in the build -- during the Cleanup Databases job on Windows.

@MattGal do the BuildPool.Windows.10.Amd64.VS2017[.Open] and BuildPool.Server.Amd64.VS2017[.Open] pools differ beyond the underlying OS. Looks like the errors below are consistent across our branches:
```
"C:\windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'F:\workspace\_work\_temp\b3ad6bac-193d-4019-9e6c-d233d9795528.ps1'"
##[error]SqlLocalDB : Stop of LocalDB instance "MSSQLLocalDB" failed because of the following error:
##[error]At F:\workspace\_work\_temp\b3ad6bac-193d-4019-9e6c-d233d9795528.ps1:2 char:1
+ SqlLocalDB stop

##[error]+ ~~~~~~~~~~~~~~~

##[error]    + CategoryInfo          : NotSpecified: (Stop of LocalDB...ollowing error::String) [], RemoteException

##[error]    + FullyQualifiedErrorId : NativeCommandError

##[error] 

##[error]PowerShell exited with code '1'.
Finishing: Cleanup databases
```

**Comment by MattGal** [comment]
> Failing builds are kind-of ugly and happen way early in the build -- during the Cleanup Databases job on Windows.
> 
> @MattGal do the BuildPool.Windows.10.Amd64.VS2017[.Open] and BuildPool.Server.Amd64.VS2017[.Open] pools differ beyond the underlying OS. Looks like the errors below are consistent across our branches:
> 
> ```
> "C:\windows\System32\WindowsPowerShell\v1.0\powershell.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'F:\workspace\_work\_temp\b3ad6bac-193d-4019-9e6c-d233d9795528.ps1'"
> ##[error]SqlLocalDB : Stop of LocalDB instance "MSSQLLocalDB" failed because of the following error:
> ##[error]At F:\workspace\_work\_temp\b3ad6bac-193d-4019-9e6c-d233d9795528.ps1:2 char:1
> + SqlLocalDB stop
> 
> ##[error]+ ~~~~~~~~~~~~~~~
> 
> ##[error]    + CategoryInfo          : NotSpecified: (Stop of LocalDB...ollowing error::String) [], RemoteException
> 
> ##[error]    + FullyQualifiedErrorId : NativeCommandError
> 
> ##[error] 
> 
> ##[error]PowerShell exited with code '1'.
> Finishing: Cleanup databases
> ```

I can't quite tell what's going on in that error, but the goal is to be all the same stuff on a Server SKU.  @ChulHul is the SME for this question.   It's increasingly clear we won't be getting rid of the Client SKU OSes any time soon, so I'd just use .Server as much as you can for the faster machine provisioning but you don't have to commit to it forever.

**Comment by dougbu** [comment]
>  the goal is to be all the same stuff on a Server SKU &hellip; I'd just use .Server as much as you can

I'm having trouble parsing your comment @MattGal. Should I wait for @ChulHul or just leave our EF6 pipelines alone i.e. close these PRs. (The _only_ change I'm making in #1610 through #1612 is from Windows 10 to Server agents.)

**Comment by MattGal** [comment]
> > the goal is to be all the same stuff on a Server SKU … I'd just use .Server as much as you can
> 
> I'm having trouble parsing your comment @MattGal. Should I wait for @ChulHul or just leave our EF6 pipelines alone i.e. close these PRs. (The _only_ change I'm making in #1610 through #1612 is from Windows 10 to Server agents.)

I didn't actually look at your pipeline logs; I was hoping @ChulHul would chime in.  My point was that we try to do the same thing in both places, but it's becoming ever more unlikely we'll go to a world where every machine is a Server SKU just because of the pain points we've had with folks moving voluntarily.  So, if you don't feel like working through the pain, you can stay on Client.

That said, I will now go look at the pipeline logs and see if I have any useful suggestions.

**Comment by MattGal** [comment]
@dougbu either SQLLocalDB just works differently on server, or you're getting extremely lucky with this working for you on the other machines.

I stood up a machine with the Server VS 2017 image we use, and aside from needing some Windows Update it was fine.  

What I did note though, is that [this script](https://github.com/dotnet/ef6/blob/fa19b658a853ac69085aa3ddc58bdf219450e6cf/azure-pipelines.yml#L73-L78) seemingly can't work more than once on such a machine and could really use some refinement.

Specifically, the mdf / ldf files it's trying to delete are actually in `%USERPROFILE%\AppData\Local\Microsoft\Microsoft SQL Server Local DB\Instances\MSSQLLocalDB\` , so a non-recursive remove-item on them is going to have no effect.  

Subsequent invocations of the same thing fail with:
```
PS C:\scratch>                   SqlLocalDB create
Creation of LocalDB instance "MSSQLLocalDB" failed because of the following error:
Unexpected error occurred inside a LocalDB instance API method call. See the Windows Application event log for error details.
```
... where said error log contains this:
```
Windows API call DeleteFileW returned error code: 32. Windows system error message is: The process cannot access the file because it is being used by another process.
Reported at line: 2496. 
```
... which makes sense, someone or some thing has a non-zero chance of having a handle open to those files when it happens. 

While I want to stress that this is only the smallest of improvements (should probably use registry or other commands to decide if stop/delete are necessary, and do a recursive search for *.*df where you're looking, I was able to confirm simply running this PS1 can work indefintely on the same machine without trouble:

``` ps1
SqlLocalDB stop
SqlLocalDB delete
Remove-Item "$env:USERPROFILE\AppData\Local\Microsoft\Microsoft SQL Server Local DB\Instances\MSSQLLocalDB\*.mdf"
Remove-Item "$env:USERPROFILE\AppData\Local\Microsoft\Microsoft SQL Server Local DB\Instances\MSSQLLocalDB\*.ldf"
SqlLocalDB create
SqlLocalDB start
```
Basically the crux of the problem is not actually deleting the db files.

**Comment by dougbu** [comment]
Thanks for the debugging @MattGal ❕ 

@bricelam what do you think of the more specific and exact script @MattGal proposed instead of https://github.com/dotnet/ef6/blob/369fa7ad1842dcf97a7969231ac05fa875d5e89d/azure-pipelines.yml#L73-L78 ❔

**Comment by dougbu** [comment]
> the mdf / ldf files it's trying to delete are actually in &hellip;

On second thought, I doubt this assertion. I believe the intent is to allow user databases e.g. Northwind to be (re)created from scratch. The `SqlLocalDB` commands alone clean up the system databases, rewriting e.g. master.mdf and tempdb.mdf. And, it's the initial `SqlLocalDB stop` command that's failing. Maybe that's because the `MSSQLLocalDB` auto-instance doesn't yet exist?

So, making the entire script conditional on user databases existing might work e.g.
``` powershell
if (Test-Path "$env:USERPROFILE\*.mdf") {
  SqlLocalDB stop
  SqlLocalDB delete
  Remove-Item "$env:USERPROFILE\*.mdf"
  Remove-Item "$env:USERPROFILE\*.ldf"
  SqlLocalDB create
  SqlLocalDB start
}
```

Thoughts @AndriySvyryd (think you wrote this part of the YAML), @ajcvickers, @bricelam❔

**Comment by AndriySvyryd** [comment]
> So, making the entire script conditional on user databases existing might work e.g.

The reason for this script to exist is that other tests were creating databases in work folders that were deleted when the agent was cleaned up leaving the instance in a corrupted state. So the condition to decide whether the instance should be recreated is whether there are any databases missing the backing files. The simplest way to do that that I can think of is dropping all databases, and if any drop fails then the instance is corrupted.

Alternatively this script could be run when the work folder containing the databases is deleted, but I don't know how to do that.

**Comment by dougbu** [comment]
Another approach might be to run the script whenever any databases exist in any work folder e.g.
```
Get-ChildItem -recurse -depth 5 $env:AGENT_BUILDDIRECTORY/../*.mdf |ForEach-Object {
  # Recreate MSSQLLocalDB
  SqlLocalDB stop
  SqlLocalDB delete
  # Does it matter much whether some database files remain in the top-level $env:USERPROFILE folder?
  SqlLocalDB create
  SqlLocalDB start
  # done
  break
}
```

Likely a waste of time to go beyond some depth. But, I chose `-depth 5` totally arbitrarily.

**Comment by AndriySvyryd** [comment]
> Another approach might be to run the script whenever any databases exist in any work folder e.g.

This isn't reliable as they might have already been deleted by a previous job on the same agent

**Comment by MattGal** [comment]
> Another approach might be to run the script whenever any databases exist in any work folder e.g.
> 
> ```
> Get-ChildItem -recurse -depth 5 $env:AGENT_BUILDDIRECTORY/../*.mdf |ForEach-Object {
>   # Recreate MSSQLLocalDB
>   SqlLocalDB stop
>   SqlLocalDB delete
>   # Does it matter much whether some database files remain in the top-level $env:USERPROFILE folder?
>   SqlLocalDB create
>   SqlLocalDB start
>   # done
>   break
> }
> ```
> 
> Likely a waste of time to go beyond some depth. But, I chose `-depth 5` totally arbitrarily.

I don't think any of the proposed scripts will help, I explicitly checked on an instance of the exact machine you're trying to build on, and running SqlLocalDB create creates files in ```"$env:USERPROFILE\AppData\Local\Microsoft\Microsoft SQL Server Local DB\Instances\MSSQLLocalDB\```, which none of these approaches seem to clean up.

**Comment by dougbu** [comment]
> Or perhaps adding SqlLocalDB start before SqlLocalDB stop fixes the error.

I doubt it because the error doesn't match what the builds log e.g.
```
> SqlLocalDB.exe stop
LocalDB instance "MSSQLLocalDB" stopped.
[servicing-flow ↓3 ↑8] C:\dd\dnx\aspnetcore-internal
> SqlLocalDB.exe delete
LocalDB instance "MSSQLLocalDB" deleted.
[servicing-flow ↓3 ↑8] C:\dd\dnx\aspnetcore-internal
> SqlLocalDB.exe stop
Stop of LocalDB instance "MSSQLLocalDB" failed because of the following error:
The specified LocalDB instance does not exist.
```

> This isn't reliable as they might have already been deleted by a previous job on the same agent

Good point

> running SqlLocalDB create creates files in `$env:USERPROFILE\AppData\Local\Microsoft\Microsoft SQL Server Local DB\Instances\MSSQLLocalDB\`, which none of these approaches seem to clean up.

Is that really a problem? The script is trying to make it possible to use SQL Local dB i.e. correct corruption cleaning workspaces causes, not clean up old log files et cetera.

**Comment by dougbu** [comment]
Another suggestion based on our TeamCity database cleanup:
``` powershell
# Drop all databases (except system and sample ones) on the current server.
$script = @'
DECLARE @name nvarchar(255)

DECLARE db CURSOR FOR
SELECT Name FROM sysdatabases
WHERE Name NOT IN ('master', 'tempdb', 'model', 'msdb')

OPEN db;

FETCH NEXT FROM db
INTO @name;

WHILE @@FETCH_STATUS = 0
BEGIN
    SET @name = REPLACE( @name, ']', ']]' )
    PRINT 'Dropping database [' + @name + ']'

    SET @name = 'DROP DATABASE [' + @name + ']'
    EXECUTE (@name)

    FETCH NEXT FROM db
    INTO @name
END

CLOSE db;
DEALLOCATE db;
'@

$env:Path="$env:Path;C:\Program Files\Microsoft SQL Server\110\Tools\Binn"
$script |sqlcmd -S '(localdb)\MSSQLLocalDB' -d master
```

@MattGal does `sqlcmd` exist on these agents? Is it already in the path?

**Comment by ajcvickers** [comment]
Isn't the real issue here that we should be able to start each test run with a known good state. The fact that SQL Server cannot be relied on to be stable if its database creation and deletion is not done correctly makes it an inherently bad choice for an image that is going to be reused. This is why we have code that attempts to get SQL Server back into a good state before we start. We really shouldn't need to do this at all.

This is one of the reasons we keep coming back to being able to run SQL Server in a container, or some other similar mechanism. We filed https://github.com/dotnet/core-eng/issues/9066 but we're not getting much traction.

**Comment by MattGal** [comment]
> Is that really a problem? The script is trying to make it possible to use SQL Local dB i.e. correct corruption cleaning workspaces causes, not clean up old log files et cetera.

I got on a machine and ran your script in a loop.  Without cleaning up these files, I got the errors I pasted for you. If this script not working more than once on the same machine is a problem for you, **yes** this is a problem.  As I've convinced myself thoroughly this has nothing to do with the Server-SKU build machines I'll let you do whatever you like.

**Comment by ErikEJ** [comment]
Unsure if this is helpful, but I am using SQL LocalDb with Azure DevOps Microsoft hosted agents. I use a library, that creates (and removes) a "temporary" SQL LocalDB instance, so it does not rely on the default instance which is named "MsSqlLocalDb". I addition, I use the SQL Srver 2017 version of localdb that allows me to use the SQL Graph  features (may not be relevant for you). My solution also works fine on my developer machine.

Happy to share more details (code / .yml scripts) if interested.

**Comment by dougbu** [comment]
@ErikEJ thanks for the suggestions but switching to a unique LocalDb instance would be a pretty bit change in this repo. And, personally, I'd rather stick with scripts that are visible in the YAML

@AndriySvyryd @ajcvickers I'm wondering if the following would be more bulletproof? It cleans up most centrally-located databases and removes any references to `*.?df` files in workspaces that have or will soon be cleaned up.
``` powershell
if (Test-Path "$env:USERPROFILE\AppData\Local\Microsoft\Microsoft SQL Server Local DB\Instances\MSSQLLocalDB") {
  SqlLocalDB stop
  SqlLocalDB delete
  if (Test-Path "$env:USERPROFILE\*.?df) {
    Remove-Item "$env:USERPROFILE\*.?df
  }
  Remove-Item "$env:USERPROFILE\AppData\Local\Microsoft\Microsoft SQL Server Local DB\Instances\MSSQLLocalDB\*.?df"
  SqlLocalDB create
  SqlLocalDB start
}
```

If we really wanted to tighten this up, would iterate through all Local dB instances e.g.
``` powershell
 dir "$env:USERPROFILE\AppData\Local\Microsoft\Microsoft SQL Server Local DB\Instances\*" -dir |
  Foreach-Object Name |
  Foreach-Object {
    SqlLocalDB stop $_
    SqlLocalDB delete $_
    Remove-Item "$env:USERPROFILE\AppData\Local\Microsoft\Microsoft SQL Server Local DB\Instances\$_\*.?df"
  }
if (Test-Path "$env:USERPROFILE\*.?df) {
  Remove-Item "$env:USERPROFILE\*.?df
}
SqlLocalDB create
SqlLocalDB start
```

For extra credit, could change the instance cleanup to
``` powershell
Remove-Item -Force -Recurse "$env:USERPROFILE\AppData\Local\Microsoft\Microsoft SQL Server Local DB\Instances\$_"
```

**Comment by ajcvickers** [comment]
@dougbu It feels like this is the wrong place to do this work. Based on, "It's increasingly clear we won't be getting rid of the Client SKU OSes any time soon" I think we should instead follow up on a reliable way to do SQL Server testing in Helix and not hack away at this.

**Comment by AndriySvyryd** [comment]
I sent a PR with a script that doesn't require instance manipulation: https://github.com/dotnet/efcore/pull/20243

**Comment by dougbu** [comment]
@AndriySvyryd last we chatted in dotnet/efcore#20243, I thought you were going to port those changes into this repo? Is something blocking that?

**Comment by AndriySvyryd** [comment]
The changes might not be needed, depending on the outcome of https://github.com/dotnet/core-eng/issues/9066

**Comment by dougbu** [comment]
@AndriySvyryd I'd like to get this PR and #1611 in but they remain blocked on getting your EF Core fixes into this repo. How does dotnet/core-eng#9066 apply here?

**Comment by dougbu** [comment]
No need for this because #1633 brings the changes in.
